CMakeLists.txt

```diff
@@ -350,13 +351,35 @@ target_link_libraries(core_interface INTERFACE
   Threads::Threads
 )
 
+# Define "sanitize_interface" with flags intended to apply to all libraries and
+# executables to support sanitizers and fuzzing, and a separate
+# "fuzzer_interface" with flags specifically intended to apply to the fuzz test
+# binary. Reason these are separate interfaces is that if -DSANITIZERS=fuzzer is
+# specified, the fuzz test binary should be built with -fsanitize=fuzzer (so it
+# can use libFuzzer's main function), but libraries should be built with
+# -fsanitize=fuzzer-no-link (so they can be linked into other executables that
+# have their own main functions).
 add_library(sanitize_interface INTERFACE)
+add_library(fuzzer_interface INTERFACE)
 target_link_libraries(core_interface INTERFACE sanitize_interface)
 if(SANITIZERS)
+  # Transform list of sanitizers into -fsanitize flags, replacing "fuzzer" with
+  # "fuzzer-no-link" in sanitize_interface flags, and moving "fuzzer" to
+  # fuzzer_interface flags.
+  string(REGEX REPLACE "(^|,)fuzzer($|,)" "\\1fuzzer-no-link\\2" SANITIZE_FLAG "${SANITIZERS}")
+  set(FUZZ_FLAG "")
+  if (NOT "${SANITIZE_FLAG}" STREQUAL "${SANITIZERS}")
+      set(FUZZ_FLAG "-fsanitize=fuzzer")
+      if (NOT BUILD_FOR_FUZZING)
+          message(FATAL_ERROR "Error: Enabling -DSANITIZERS=fuzzer without -DBUILD_FOR_FUZZING=ON is not supported.")
```

Comment:
style: Our CMake files use 2-space for indentation and no space after `if` like this: `if(FOO`. When we were switching to CMake I assumed this is ok because it was all consistent. Now I realize that people will keep adding changes that use 4-space indentation and a space after `if` because the rest of the code base uses that.

Maybe just ignore this and accept that our CMake files will end up with an inconsistent mixture of different styles?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31741#discussion_r1961422233

Good catch, fixed indentation and spacing. I don't think it should be too much of a burden to enforce a consistent style for cmake code given how little cmake code there is relatively. There also seems to be a [cmake-lint](https://cmake-format.readthedocs.io/en/latest/lint-usage.html) tool that could maybe help.

Reply:
Wow! I didn't know cmake-lint existed and didn't even imagine it could exist ;) It would be perfect to enforce this in CI.